packages/vite/src/node/plugins/asset.ts

```diff
@@ -470,15 +474,24 @@ async function fileToBuiltUrl(
       map.set(contentHash, fileName)
     }
     const emittedSet = emittedHashMap.get(config)!
+    const duplicates = duplicateAssets.get(config)!
+    const name = normalizePath(path.relative(config.root, file))
     if (!emittedSet.has(contentHash)) {
-      const name = normalizePath(path.relative(config.root, file))
       pluginContext.emitFile({
         name,
         fileName,
         type: 'asset',
         source: content
       })
       emittedSet.add(contentHash)
+    } else {
```

Comment:
I think we need to make `duplicates` a `Map` from `file` to duplicate object here, instead of an array. The `cache` uses `id` as a key, so two files that only differ on the postfix (like a SVG fragment) will end up generating two duplicates entries.
Another option is to change the cache to work at with `file` as the key and extract the logic for handling the postfix out the cache guard. Probably the map is easier to do for this PR, but the latter may be better in the long run.

Reply:
I refactored to a map as I wasn't sure about the cache stuff sorry!